src/ILLink.RoslynAnalyzer/DynamicallyAccessedMembersAnalyzer.cs

```diff
@@ -0,0 +1,683 @@
+		}
+
+		static void ProcessReturnStatement (ReturnStatementSyntax returnStatement, SyntaxNodeAnalysisContext context)
+		{
+			if (returnStatement.Expression == null || context.SemanticModel.GetSymbol (returnStatement.Expression) is not ISymbol returnValueSymbol ||
+				returnValueSymbol.TryGetDynamicallyAccessedMemberTypes (out var damtOnReturnValue))
+				return;
+
+			bool returnValueIsAMethodCall = returnValueSymbol is IMethodSymbol;
+			if (returnStatement.Expression is ThisExpressionSyntax)
+				returnValueSymbol = returnValueSymbol.ContainingSymbol;
+
+			var containingMethod = returnStatement.Ancestors ().OfType<MethodDeclarationSyntax> ().First ();
+			if (context.SemanticModel.GetDeclaredSymbol (containingMethod) is IMethodSymbol containingMethodSymbol &&
+				containingMethodSymbol.TryGetDynamicallyAccessedMemberTypesOnReturnType (out var damtOnReturnType) &&
+				!SourceHasMatchingAnnotations (damtOnReturnValue, damtOnReturnType, out var missingAnnotations))
+					ReportUnmatchedAnnotations (returnStatement.GetLocation (),
+						returnValueSymbol, containingMethodSymbol, missingAnnotations, context, sourceIsMethodReturnType: returnValueIsAMethodCall);
+		}
+
+		static bool SourceHasMatchingAnnotations (DynamicallyAccessedMemberTypes? sourceMemberTypes, DynamicallyAccessedMemberTypes? targetMemberTypes, out string missingMemberTypesString)
+		{
+			missingMemberTypesString = $"'{nameof (DynamicallyAccessedMemberTypes.All)}'";
+			if (targetMemberTypes == null)
+				return true;
+
+			sourceMemberTypes ??= DynamicallyAccessedMemberTypes.None;
+			var missingMemberTypesList = Enum.GetValues (typeof (DynamicallyAccessedMemberTypes))
+				.Cast<DynamicallyAccessedMemberTypes> ()
+				.Where (damt => (damt & targetMemberTypes & ~sourceMemberTypes) == damt && damt != DynamicallyAccessedMemberTypes.None)
+				.ToList ();
+
+			if (missingMemberTypesList.Count == 0)
+				return true;
+
+			if (missingMemberTypesList.Contains (DynamicallyAccessedMemberTypes.PublicConstructors) &&
+				missingMemberTypesList.SingleOrDefault (mt => mt == DynamicallyAccessedMemberTypes.PublicParameterlessConstructor) is var ppc &&
+				ppc != DynamicallyAccessedMemberTypes.None)
+				missingMemberTypesList.Remove (ppc);
+
+			missingMemberTypesString = string.Join (", ", missingMemberTypesList.Select (mmt => $"'DynamicallyAccessedMemberTypes.{mmt}'"));
+			return false;
+		}
+
+			static void ReportUnmatchedAnnotations (
+			Location diagnosticLocation,
+			ISymbol sourceSymbol,
+			ISymbol targetSymbol,
+			string missingMemberTypes,
+			SyntaxNodeAnalysisContext context,
+			bool sourceIsMethodReturnType = false)
+		{
+			Debug.Assert (!string.IsNullOrEmpty (missingMemberTypes));
+
+			switch ((sourceSymbol, targetSymbol)) {
+			case (IParameterSymbol sourceParameter, IParameterSymbol targetParameter):
+				context.ReportDiagnostic (Diagnostic.Create (
+					s_IL2067,
```

Comment:
Is the only difference in many of these diagnostics that the source is one type of symbol and the target is another type of symbol?

This seems like a difficult design. As a user this would mean that if I want to silence all the conversion warnings in a method I would have to individually silence each warning code, but I don't see what value I get from having a specific warning code for each one.

These all feel like instances of the same warning.

Reply:
At minimum, we should try to extract some of this code into a helper, but really I think we should try to figure out why we need this many warning codes.

Reply:
Ah I'm looking at the messages now -- some of them are significantly different. We could probably share more but I see why it's not obvious how to do so. Probably a good investment though.

Reply:
We have a helper in linker to produce the right warning.

The many different warning codes were introduce actually according to your (@agocke) guidance that each unique message should have its own warning code. It's true that silencing them all is tricky, but I see that as a benefit (you should not want to do that manually, have the `SuppressTrimAnalysisWarnings` for it) - the simplest way to silence them in code is to add `RequiresUnreferencedCode` onto the method which produces them.